custom_metrics/sass.js

```diff
@@ -0,0 +1,308 @@
+	});
+	ret.variables = sortObject(ret.variables);
+
+	// Mixins
+	// Note: We are not taking the Indented Mixin Syntax into account
+	ret.mixins = {};
+	scss.replace(/@mixin\s+([\w-]+)(?:\((.+?)\))?/g, ($0, name, args) => {
+		// Note: - and _ are equivalent in mixin names and refer to the same mixin
+		//         we are not taking that into account here. This applies to functions too (below)
+		ret.mixins[name] = {
+			// Default values can't have commas, so split should work
+			args: args?.split(/\s*,\s*/), // TODO separate name from default value?
+			calls: scss.match(RegExp("@include\\s+" + name + "\\b", "gi"))?.length
+		};
+	});
+
+	// Custom functions
+	ret.functions = {};
+	let lastName; // used to track down the nearest @return after a @function
+	scss.replace(/@function\s+([\w-]+)(?:\((.+?)\))?|@return (.+)/g, ($0, name, args, returnValue) => {
+		if ($0.indexOf("@return") === 0 && lastName) {
+			ret.functions[lastName].return = returnValue;
+			lastName = undefined;
+			return $0;
+		}
+
+		lastName = name;
+		ret.functions[name] = {
+			// Default values can't have commas, so split should work
+			args: args?.split(/\s*,\s*/), // TODO separate name from default value?
+			calls: scss.match(RegExp("\\b" + name + "\\(", "gi"))?.length
+		};
+	});
+
+	// TODO Measure usage of Sass functions
+	ret.functionCalls = {};
+	let functionCallRegex = new RegExp(`(?<![-.\\w])(${SassFunctions.join("|").replace(".", "\\.")})\\(`, "gi");
+	scss.replace(functionCallRegex, ($0, name) => {
+		ret.functionCalls[name] = (ret.functionCalls[name] || 0) + 1;
+	});
+	ret.functionCalls = sortObject(ret.functionCalls);
+
+	// Conditionals
+	ret.ifs = [];
+	scss.replace(/@(?:(else )?if\s+?(.+?)|@else)(?=\s*\{)/g, ($0, elseIf, test) => {
+		if ($0 === "@else") {
+			// Else for previous condition
+			let obj = ret.ifs.pop();
+			console.log(ret.ifs, obj);
+			obj.hasElse = true;
+			ret.ifs.push(obj);
+		}
+		else if (elseIf) {
+			// Else if for previous condition
+			let obj = ret.ifs.pop();
+			obj.elseIfs = obj.elseIfs || [];
+			obj.elseIfs.push(test);
+			ret.ifs.push(obj);
+		}
+		else {
+			// New conditional
+			ret.ifs.push({test});
+		}
+	});
+
+	// @each, @for, @while Loops
+	ret.eaches = {};
+	scss.replace(/@each\s+(.+ in .+)(?=\s*\{)/g, ($0, args) => {
+		ret.eaches[args] = (ret.eaches[args] || 0) + 1;
+	});
+
+	ret.fors = {};
+	// console.log("@for", scss);
+	scss.replace(/@for\s+(.+)(?=\s*\{)/g, ($0, args) => {
+		ret.fors[args] = (ret.fors[args] || 0) + 1;
+	});
+
+	ret.whiles = {};
+	scss.replace(/@while\s+(.+)(?=\s*\{)/g, ($0, args) => {
+		ret.whiles[args] = (ret.whiles[args] || 0) + 1;
+	});
+
+	// @extend
+	ret.extends = {};
+	scss.replace(/@extend\s*(.+)(?=\s*(?:[;}]|$))/g, ($0, selector) => {
+		ret.extends[selector] = (ret.extends[selector] || 0) + 1;
+	});
+
+	// @error
+	ret.errors = scss.match(/@error (.+)/g)?.map(e => e.slice(7));
+
+	// CSS variables that are set with Sass variables
```

Comment:
This may be over-specific, or need additional tests:

- custom props defined using other sass: `--<var>: #{<anything>};` (We often use functions here rather than variables)
- custom props **named by** sass: `--#{<var>}: <anything>;`

Might be interesting to look for the `#{}` interpolation pattern other places as well. It can be used for defining selectors, property names, etc – often inside mixins or loops.

Reply:
> As far as I can tell, the .sass syntax is far less popular, but does have a dedicated niche following. It would be interesting to know how much of a following that adds up to…

The code does count all extensions, so it will still give us info on the relative popularity of scss vs sass, it just won't analyze the sass files. 

> Are we capturing and comparing file/path names to compare how sass files are structured? It would be interesting to know if there are common file/directory naming structures, or how many people are using Sass loaded from npm modules…

Any ideas what we can store for this that is sufficiently compact?

> This may be over-specific, or need additional tests:

Oh good point, I even saw this in the Bootstrap SCSS files but wasn't sure it's common enough to study.
Selectors are *really* hard to track down without a parser I'm afraid :(

Reply:
Alright, I just pushed a heuristic to measure CSS variables whose name is set via Sass interpolation.